File: cscm12-2324/Countdown.java
import java.lang.Comparable;
import java.util.*;


enum Op { Plus, Minus, Times, Div}

class ArithmeticExpression implements Comparable<ArithmeticExpression> {


  private final ArithmeticExpression subLeft;  // the left subexpression
  private final ArithmeticExpression subRight; // the right subexpression
  private final Op op; // the top-level operator

  private final int val; // the integer value; only useful if this is a leaf

  private final int hashCode; // the hash of the object

  private ArithmeticExpression(int i)
  {
    subLeft = subRight = null;
    op = null;
    val = i;
    hashCode = scramble(i);
  }

  private ArithmeticExpression(ArithmeticExpression l, Op iop, ArithmeticExpression r)
  {
    subLeft = l;
    subRight = r;
    op = iop;
    val = 0;
    int opint = 8392;
    switch(op)
    {
      case Plus:
        opint = 0;
        break;
      case Minus:
        opint = 1;
        break;
      case Times:
        opint = 2;
        break;
      case Div:
        opint = 3;
        break;
    }
    hashCode = pair(l.hashCode, pair(opint, r.hashCode));
  }

  private int pair(int x, int y)
  {
    return scramble((y + (x + y) * (x + y + 1))/2);
  }

  private int scramble(int x)
  {
    //taken from https://stackoverflow.com/questions/664014/what-integer-hash-function-are-good-that-accepts-an-integer-hash-key
    x = ((x >> 16) ^ x) * 0x45d9f3b;
    x = ((x >> 16) ^ x) * 0x45d9f3b;
    x = (x >> 16) ^ x;
    return x;
  }

  public boolean equals(ArithmeticExpression e)
  {
    if(hashCode() != e.hashCode())
      return false;
    if(op == null)
      return e.op == null && val == e.val;
    else
      return e.op == op && subLeft.equals(e.subLeft) && subRight.equals(e.subRight);
  }

  private static int compareOp(Op o, Op oo)
  {
    switch(o)
    {
      case Plus:
        return (oo == Op.Plus) ? 0 : 1;
      case Times:
        switch(oo)
        {
          case Plus:
            return -1;
          case Times:
            return 0;
          default:
            return 1;
        }
      case Minus:
        switch(oo)
        {
          case Minus:
            return 0;
          case Div:
            return 1;
          default:
            return -1;
        }
      case Div:
        return (oo == Op.Div) ? 0 : -1;
    }
    return 111;
  }

  public int compareTo(ArithmeticExpression e)
  {
    if(hashCode() != e.hashCode())
      return (hashCode() < e.hashCode()) ? 1 : -1;
    if(op == null)
      return (e.op == null) ? 0 : 1; 
    if(e.op == null)
      return -1;
    int os = compareOp(op, e.op);
    if(os != 0)
      return os;
    int ls = subLeft.compareTo(e.subLeft);
    if(ls != 0)
      return ls;
    return subRight.compareTo(e.subRight);
  }

  public int hashCode()
  {
    return hashCode;
  }

  private static String paren(String s)
  {
    return "(" + s + ")";
  }

  public String toString()
  {
    if(op == null)
      return "" + val;
    String res = "";
    res += paren(subLeft.toString());
    switch(op)
    {
      case Plus:
        res += " + ";
        break;
      case Minus:
        res += " - ";
        break;
      case Times:
        res += " * ";
        break;
      case Div:
        res += " / ";
        break;
    }
    res += paren(subRight.toString());
    return res;
  }

}

class CountdownSol {

  // A hash table meant to map a list of numbers to all possible arithmetic
  // expressions that can be generated from it
  static Hashtable<LinkedList<Integer>,TreeSet<ArithmeticExpression>>
      memoAllExpressions = new Hashtable<LinkedList<Integer>,TreeSet<ArithmeticExpression>>();


  public static ArithmeticExpression countdown(LinkedList<Integer> numbers, int goal)
  { 
    // your code goes here
    return null;
  }

  public static LinkedList<LinkedList<Integer>> allSplits(LinkedList<Integer> xs)
  {
    LinkedList<LinkedList<Integer>> res = new LinkedList<LinkedList<Integer>>();
    if(xs.size() == 2)
    {
      LinkedList<Integer> xsc = new LinkedList<Integer>(xs);
      LinkedList<Integer> e0 = new LinkedList<Integer>();
      LinkedList<Integer> e1 = new LinkedList<Integer>();
      e0.push(xsc.pop());
      e1.push(xsc.pop());
      res.push(e0);
      res.push(e1);
    } 
    else if(xs.size() > 2)
    {
      int k = xs.pop();
      LinkedList<LinkedList<Integer>> yss = allSplits(xs);
      while(!yss.isEmpty())
      {
        LinkedList<Integer> ys0 = yss.pop();
        LinkedList<Integer> ys1 = yss.pop();
        LinkedList<Integer> ys0copy = new LinkedList<Integer>(ys0);
        LinkedList<Integer> ys1copy = new LinkedList<Integer>(ys1);
        ys0.push(k);
        ys1copy.push(k);
        res.push(ys0);
        res.push(ys1);
        res.push(ys0copy);
        res.push(ys1copy);
      }
      xs.push(k);
    }
    return res;
  }

  public static void playCountdown(LinkedList<Integer> countdowninstance, int goal)
  {
    ArithmeticExpression sol = countdown(countdowninstance, goal);
    if(sol == null)
      System.out.println("it is not possible to reach " + goal + " using " + countdowninstance);
    else
    {
      System.out.println(goal + " can be reached using " + countdowninstance);
      System.out.println("Here is the an arithmetic expression witnessing that\n");
      System.out.println(sol);   
    }
  }

  public static void main(String[] arr)
  {
    LinkedList<Integer> xs = new LinkedList<Integer>();
    for(int i = 0; i < 6; ++i)
      xs.push(i);
    System.out.println(xs);
    System.out.println(allSplits(xs));

    LinkedList<Integer> countdowninstance = new LinkedList<Integer>();
    countdowninstance.push(5);
    countdowninstance.push(5);
    countdowninstance.push(4);
    countdowninstance.push(3);
    countdowninstance.push(2);

    playCountdown(countdowninstance, 6);

    playCountdown(xs, 876);
  }
}

File: cscm712-2526/countdown.py
from copy import copy

Ops = '+*/-'
OpsFun = { '+' : lambda x, y: x.__add__(y),\
           '*' : lambda x, y: x.__mul__(y),\
           '/' : lambda x, y: x.__floordiv__(y),\
           '-' : lambda x, y: x.__sub__(y)\
         }
maxOpCode = max([ ord(i) for i in Ops])

class ArithmeticExpression:
  """
  We want the following attributes around:
  * subLeft for the left subexpression
  * subRight for the right subexpression
  * label for the top-level operator or the integer value of the expression
    The operators '+*/-' will be stored as strings, the integers as ints.
  * hashCode, for the hash of the whole thing
  * value, for the value the expression evaluates to

  This datastructure is meant to be immutable; don't mess with the attributes
  from outside of the class
  """

  def __init__(self, x, l = None, r = None):
      self.label = x
      self.subLeft = l
      self.subRight = r
      if type(x) == int:
          x += 1 + maxOpCode
      self.hashCode = hash((x, l, r))
      if self.subLeft == None:
          self.value = self.label
      else:
          if self.subLeft.value == None or self.subRight.value == None or\
             self.label == '/' and\
             (self.subRight.value == 0 or self.subLeft.value % self.subRight.value != 0):
              self.value = None
          else:
              self.value = OpsFun[self.label](self.subLeft.value, self.subRight.value)

  def eval(self):
      return self.value

  def __hash__(self):
      return self.hashCode

  def __eq__(self, other):
      return other != None and\
             self.hashCode == other.hashCode and\
             self.label == other.label and\
             self.subRight == other.subRight and\
             self.subLeft == other.subLeft

  # not terribly efficient
  def __str__(self):
      if self.subLeft == None:
          return str(self.label)
      else:
          return '(' + str(self.subLeft) + ' ' + self.label + ' ' + str(self.subRight) + ')'
  



# A helper function you may want to use!
def allSplits(xs):
  def aux(xs):
      res = []
      if len(xs) == 2:
         res = [([xs[0]], [xs[1]])]
      elif len(xs) > 2:
         k = xs.pop()
         yss = aux(xs)
         while len(yss) > 0:
             (ys0, ys1) = yss.pop()
             ys0copy = copy(ys0)
             ys1copy = copy(ys1)
             ys0.append(k)
             ys1copy.append(k)
             for i in [(ys0, ys1), (ys0copy, ys1copy)]:
                 res.append(i)
         xs.append(k)
      return res
  return list(map(lambda x: tuple(map(tuple, x)), aux(list(xs))))


"""
TODO: the countdown function takes as input a tuple of integers and an integer,
and should return an arithmetic expression that uses integers in the tuple
at most once to computer the second integer.

You should define the auxiliary function populateMemo whose purpose is to
fill the hash table memoAllExpressions. populateMemo(numbers) should fill
all cells indexed by sublists of numbers.
"""

# A hash table meant to map a list of numbers to a set of all possible arithmetic
# expressions that can be generated from it; here I give you some examples in
# the initialization
memoAllExpressions = { (0,) : {ArithmeticExpression(0)},\
                       (2, 4) : { ArithmeticExpression(4),\
                                  ArithmeticExpression(2),\
                                  ArithmeticExpression('+', ArithmeticExpression(2), ArithmeticExpression(4)),\
                                  ArithmeticExpression('-', ArithmeticExpression(4), ArithmeticExpression(2)),\
                                  ArithmeticExpression('*', ArithmeticExpression(2), ArithmeticExpression(4)),\
                                  ArithmeticExpression('/', ArithmeticExpression(4), ArithmeticExpression(2))}\
                    }

def populateMemo(numbers):
  return
def countdown(numbers, goal):
  return

def playCountdown(numbers, goal):
  sol = countdown(countdowninstance, goal)
  if sol == None:
      print('it is not possible to reach', goal, 'using', countdowninstance)
  else:
      print(goal, 'can be reached using', countdowninstance)
      print('Here is the an arithmetic expression witnessing that')
      print(sol)

if __name__ == "__main__":
  ys = (3, 2, 1, 0)
  print(allSplits(ys))
  xs = tuple(range(0,6))
  print(xs)
  print(allSplits(xs))
  countdowninstance = (5,4,3,2)
  playCountdown(countdowninstance, 6)
